compiler/x/codegen/OMRTreeEvaluator.cpp

```diff
@@ -4779,6 +4779,148 @@ TR::Register *OMR::X86::TreeEvaluator::atomicorEvaluator(TR::Node *node, TR::Cod
+      List<TR::Register> popRegisters(cg->trMemory());
+      persistentConditions = generateRegisterDependencyConditions(GRANode, cg, 0, &popRegisters);
+      cg->decReferenceCount(GRANode);
+      }
+
+   if (transientFailureNode->getNumChildren() != 0)
+      {
+      GRANode = transientFailureNode->getFirstChild();
+      cg->evaluate(GRANode);
+      List<TR::Register> popRegisters(cg->trMemory());
+      transientConditions = generateRegisterDependencyConditions(GRANode, cg, 0, &popRegisters);
+      cg->decReferenceCount(GRANode);
+      }
+
+   //startLabel
+   //add place holder register so that eax would not contain any useful value before xbegin
+   TR::Register *dummyReg = cg->allocateRegister();
+   dummyReg->setPlaceholderReg();
+   TR::RegisterDependencyConditions *startLabelConditions = generateRegisterDependencyConditions((uint8_t)0, 1, cg);
+   startLabelConditions->addPostCondition(dummyReg, TR::RealRegister::eax, cg);
+   startLabelConditions->stopAddingConditions();
+   cg->stopUsingRegister(dummyReg);
+   generateLabelInstruction(LABEL, node, startLabel, startLabelConditions, cg);
+
+   //xbegin, if fallback then go to fallbackLabel
+   generateLongLabelInstruction(XBEGIN4, node, fallbackLabel, cg);  
+
+   //jump to  fallThrough Path
+   if (fallThroughConditions)
+      generateLabelInstruction(JMP4, node, fallThroughLabel, fallThroughConditions, cg);
+   else
+      generateLabelInstruction(JMP4, node, fallThroughLabel, cg);
+
+   endLabelConditions = generateRegisterDependencyConditions((uint8_t)0, 1, cg);
+   endLabelConditions->addPostCondition(accReg, TR::RealRegister::eax, cg);
+   endLabelConditions->stopAddingConditions();
+  
+   //Label fallback begin:
+   generateLabelInstruction(LABEL, node, fallbackLabel, cg);
+
+   //test eax, 0x2, jne to transientFailure
+   generateRegImmInstruction(TEST1AccImm1, node, accReg, 0x2, cg);
+   if(transientConditions)
+      generateLabelInstruction(JNE4, node, transientFailureLabel, transientConditions, cg);
+   else
+      generateLabelInstruction(JNE4, node, transientFailureLabel, cg);
+
+   cg->stopUsingRegister(accReg);
+ 
+   //jmp to persistent begin:
+   if (persistentConditions)
+      generateLabelInstruction(JMP4, node, persistentFailureLabel, persistentConditions, cg);
+   else
+      generateLabelInstruction(JMP4, node, persistentFailureLabel, cg);
+
+   //Label finish
+   generateLabelInstruction(LABEL, node, endLabel, endLabelConditions, cg);
+
+   cg->decReferenceCount(persistentFailureNode);
+   cg->decReferenceCount(transientFailureNode);
+   
+   return NULL;
+   }
+
+TR::Register *
+OMR::X86::TreeEvaluator::tfinishEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   generateInstruction(XEND, node, cg);
+   return NULL;
+   }
+
+TR::Register *
+OMR::X86::TreeEvaluator::tabortEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   //we hardcode an abort reason as 0x04 just for simplicity.
+   generateImmInstruction(XABORT, node, 0x04, cg);
```

Comment:
This seems dangerous and ill advised - if it has to survive there had better at least be a big TODO here to fix this.